src/engraving/dom/systemlock.cpp

```diff
@@ -0,0 +1,163 @@
+/*
+ * SPDX-License-Identifier: GPL-3.0-only
+ * MuseScore-Studio-CLA-applies
+ *
+ * MuseScore Studio
+ * Music Composition & Notation
+ *
+ * Copyright (C) 2021 MuseScore Limited
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "measurebase.h"
+#include "page.h"
+#include "score.h"
+#include "system.h"
+#include "systemlock.h"
+
+#include "log.h"
+
+using namespace muse::draw;
+
+namespace mu::engraving {
+bool SystemLock::contains(const MeasureBase* mb) const
+{
+    return m_startMB->isBeforeOrEqual(mb) && mb->isBeforeOrEqual(m_endMB);
+}
+
+void SystemLocks::add(const SystemLock* lock)
+{
+    m_systemLocks.emplace(lock->startMB(), lock);
+
+#ifndef NDEBUG
+    sanityCheck();
+#endif
+}
+
+void SystemLocks::remove(const SystemLock* lock)
+{
+    m_systemLocks.erase(lock->startMB());
+
+#ifndef NDEBUG
+    sanityCheck();
+#endif
+}
+
+const SystemLock* SystemLocks::lockStartingAt(const MeasureBase* mb) const
+{
+    auto iter = m_systemLocks.find(mb);
+    return iter != m_systemLocks.end() ? (*iter).second : nullptr;
+}
+
+const SystemLock* SystemLocks::lockContaining(const MeasureBase* mb) const
+{
+    if (m_systemLocks.empty()) {
+        return nullptr;
+    }
+
+    auto iter = m_systemLocks.lower_bound(mb);
+    if (iter != m_systemLocks.begin() && (iter == m_systemLocks.end() || mb->isBefore((*iter).second->startMB()))) {

```

Comment:
```suggestion
    if (iter != m_systemLocks.begin() && (iter == m_systemLocks.end() || mb->isBefore(iter->second->startMB()))) {
```
(that could be done everywhere where `(*iter).` is used)

or maybe for readability
```suggestion
    if (iter != m_systemLocks.begin() 
        && (iter == m_systemLocks.end() || mb->isBefore(iter->second->startMB()))) {
```